

from database.models import Reference



def create_or_update_references(cve, cve_references):
    try:
        for reference_data in cve_references: 
            url=reference_data['url']
            
            existing_reference = get_existing_reference(cve, url)
            
            if existing_reference:
                update_existing_reference( existing_reference, reference_data)
            else:
                create_new_reference(cve, reference_data)

    except Exception as exc:
        raise Exception(f"Failed to create or update reference_data: {exc}")


def get_existing_reference(cve, url):
    try:
        return Reference.objects.filter(cve=cve, url=url).first()
    except Exception as e:
        raise Exception(f"Failed to get existing Reference: {e}")


def update_existing_reference(existing_reference, reference_data):
    try:
        source=reference_data['source']
        if existing_reference.source != source:
            existing_reference.source = source
            existing_reference.save()
            print("    UPDATE -> reference")

    except Exception as e:
        raise Exception(f"Failed to update existing reference: {e}")


def create_new_reference(cve, reference_data):
    try:
        Reference.objects.create(
            cve=cve,
            url=reference_data['url'],
            source=reference_data['source']
        )
        print("    CREATE -> reference")
    except Exception as e:
        raise Exception(f"Failed to create new reference: {e}")
